src/structures/suggestion.ts

```diff
@@ -0,0 +1,176 @@
+
+	constructor(protected readonly raw: SuggestionSelect) {
+		super(raw.upvotedBy?.split(","), raw.downvotedBy?.split(","));
+	}
+
+	/** Upvote the suggestion. */
+	public async upvote(userId: string): Promise<void> {
+		if (this.hasUpdatedStatus()) return;
+
+		super.upvote(userId);
+
+		const raw = await db
+			.select({
+				downvotedBy: DbSuggestion.downvotedBy,
+				upvotedBy: DbSuggestion.upvotedBy,
+			})
+			.from(DbSuggestion)
+			.where(eq(DbSuggestion.id, this.raw.id))
+			.get();
+
+		const upvotes = new Set(raw?.upvotedBy?.split(",") ?? []);
+		const downvotes = new Set(raw?.downvotedBy?.split(",") ?? []);
+
+		upvotes.add(userId);
+		downvotes.delete(userId);
+
+		const serializedUpvotes = [...upvotes].join(",") || null;
+		const serializedDownvotes = [...downvotes].join(",") || null;
+
+		await db
+			.update(DbSuggestion)
+			.set({
+				upvotedBy: serializedUpvotes,
+				downvotedBy: serializedDownvotes,
+			})
+			.where(eq(DbSuggestion.id, this.raw.id));
+
+		this.raw.upvotedBy = serializedUpvotes;
+		this.raw.downvotedBy = serializedDownvotes;
+	}
+
+	/** Downvote the suggestion. */
+	public async downvote(userId: string): Promise<void> {
+		if (this.hasUpdatedStatus()) return;
+
+		super.downvote(userId);
+
+		const raw = await db
+			.select({
+				downvotedBy: DbSuggestion.downvotedBy,
+				upvotedBy: DbSuggestion.upvotedBy,
+			})
+			.from(DbSuggestion)
+			.where(eq(DbSuggestion.id, this.raw.id))
+			.get();
+
+		const upvotes = new Set(raw?.upvotedBy?.split(",") ?? []);
+		const downvotes = new Set(raw?.downvotedBy?.split(",") ?? []);
+
+		upvotes.delete(userId);
+		downvotes.add(userId);
+
+		const serializedUpvotes = [...upvotes].join(",") || null;
+		const serializedDownvotes = [...downvotes].join(",") || null;
+
+		await db
+			.update(DbSuggestion)
+			.set({
+				upvotedBy: serializedUpvotes,
+				downvotedBy: serializedDownvotes,
+			})
+			.where(eq(DbSuggestion.id, this.raw.id));
+
+		this.raw.upvotedBy = serializedUpvotes;
+		this.raw.downvotedBy = serializedDownvotes;
+	}
+
+	/** Remove the user's vote for the suggestion. */
+	public async removeVote(userId: string): Promise<void> {
+		await db.delete(DbSuggestion).where(eq(DbSuggestion.id, this.raw.id));
+		super.removeVote(userId);
+	}
+
+	/** Set the status of the suggestion. */
+	public async setStatus(
+		user: User,
+		status: SuggestionStatus,
+		reason?: string,
+	): Promise<void> {
+		this.raw.statusUserId = user.id;
+		this.raw.status = status;
+		this.raw.statusReason = reason ?? null;
+		this.raw.updatedAt = new Date();
+
+		await db.update(DbSuggestion).set({
+			statusUserId: this.raw.statusUserId,
+			status: this.raw.status,
+			statusReason: this.raw.statusReason,
+			updatedAt: this.raw.updatedAt,
+		});
+	}
+
+	/** Type-guard that checks whether the suggestion status has been updated and is not `"POSTED"`. */
+	public hasUpdatedStatus(): boolean {
+		return this.status !== "POSTED";
+	}
+
+	get id(): number {
```

Comment:
These getters seem kind of like an unnecessary abstraction compared to just accessing properties directly.